script/modelling/logistic_regression.jl

```diff
@@ -0,0 +1,217 @@
+#' achieve a greater score in cross-validation.
+
+#' ## Reformulation as a conic optimization problem
+#' By introducing auxiliary variables $t_1, \cdots, t_n$ and $r$,
+#' the optimization problem is equivalent to
+#' $$
+#' \begin{aligned}
+#' \min_{t, r, \theta} \;& \sum_{i=1}^n t_i + \lambda r \\
+#' \text{subject to } & \quad t_i \geq \log(1 + \exp(- y_i \theta^\top x_i)) \\
+#'                    & \quad r \geq \|\theta\|
+#' \end{aligned}
+#' $$
+#' Now, the trick is to reformulate the constraints $t_i \geq \log(1 + \exp(- y_i \theta^\top x_i))$
+#' with the help of the *exponential cone*
+#' $$
+#' K_{exp} = \{ (x, y, z) \in \mathbb{R}^3 : \; y \exp(x / y) \leq z \} .
+#' $$
+#' Indeed, by passing to the exponential, we
+#' see that for all $i=1, \cdots, n$, the constraint $t_i \geq \log(1 + \exp(- y_i \theta^\top x_i))$
+#' is equivalent to
+#' $$
+#' \exp(-t_i) + \exp(u_i - t_i) \leq 1
+#' $$
+#' with $u_i = -y_i \theta^\top x_i$. Then, by adding two auxiliary variables
+#' $z_{i1}$ and $z_{i2}$ such that $z_{i1} \geq \exp(u_i-t_i)$ and $z_{i2} \geq \exp(-t_i)$, we get
+#' the equivalent formulation
+#' $$
+#' \left\{
+#' \begin{aligned}
+#' (u_i -t_i , 1, z_{i1}) & \in  K_{exp}  \\
+#' (-t_i , 1, z_{i2}) & \in  K_{exp}  \\
+#' z_{i1} + z_{i2} & \leq  1
+#' \end{aligned}
+#' \right.
+#' $$
+#' In this setting, the conic version of the logistic regression problems writes out
+#' $$
+#' \begin{aligned}
+#' \min_{t, z, r, \theta}&  \; \sum_{i=1}^n t_i + \lambda r \\
+#' \text{subject to } & \quad  (u_i -t_i , 1, z_{i1})  \in  K_{exp}  \\
+#'                    & \quad  (-t_i , 1, z_{i2})  \in  K_{exp}  \\
+#'                    & \quad  z_{i1} + z_{i2}  \leq  1 \\
+#'                    & \quad u_i = -y_i x_i^\top \theta \\
+#'                    & \quad r \geq \|\theta\|
+#' \end{aligned}
+#' $$
+#' and thus encompasses $3n + p + 1$ variables and $3n + 1$ constraints ($u_i = -y_i \theta^\top x_i$
+#' is only a temporary constraint used to clarify the notation).
+#' Thus, if $n \gg 1$, we get a large number of variables and constraints which
+#' could imped the resolution in the conic solver.
+
+#' ## Fitting logistic regression with a conic solver
+#' It is now time to pass to the implementation. We choose ECOS as a conic solver.
+using JuMP
+using Random
+using ECOS
+
+Random.seed!(2713);
+
+#' We start by implementing a function to generate a fake dataset, and where
+#' we could tune the correlation between the feature variables. The function
+#' is a direct transcription of the one used in [this blog post](http://fa.bianp.net/blog/2013/numerical-optimizers-for-logistic-regression/).
+function generate_dataset(n_samples=100, n_features=10; corr=0.0)
+    X = randn(n_samples, n_features)
+    w = randn(n_features)
+    y = sign.(X * w)
+    X .+= 0.8 * randn(n_samples, n_features) # add noise
+    X .+= corr # this makes it correlated by adding a constant term
```

Comment:
I fail to see, how does a constant term adds correlation? Is it not just moving the data points by a constant vector in the feature space ?

If we need correlated features, should randn be called with a correlation matrix?

Reply:
You are right about that, I should have been more careful when adapting the code from the origin blog post. I have renamed `corr` to `shift`, as we are only shifting the features by a constant value.